packages/react-form-state/README.md

```diff
@@ -0,0 +1,947 @@
+  }}
+```
+
+When given an [async](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function) function (or one that returns a [promise](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), the `children` function will be rerendered with `submitting: true` until the promise resolves or rejects. Once it has returned the function will be rerendered with `submitting: false` and `errors` either `[]` (if the promise resolved with no value), or with an array of `{field?: string, error: string}` if the promise resolved with an error list.
+
+#### Example
+
+```typescript
+// api.ts
+
+const fakeErrorResult = {error: [{message: 'api is not real'} };
+
+// simple async function that fakes a failing api call
+export function fakeApiCall(data: any) {
+  return new Promise(resolve => {
+    setTimeout(() => {
+      resolve(]);
+    }, 2000);
+  });
+}
+```
+
+```typescript
+// MyPage.ts
+import {fakeApiCall} from './api';
+import {TextField, Button, Banner, FormLayout, Spinner} from '@shopify/polaris';
+import FormState from '@shopify/react-form-state';
+
+export function MyPage() {
+  return (
+    <FormState
+      initialValues={{email: '', password: ''}}
+      onSubmit={async () => {
+        const results = await fakeApiCall();
+        if (results.error) {
+          return results.error;
+        }
+      }}
+    >
+      {({fields, submitting, error, submit}) => {
+        return (
+          <form onSubmit={submit}>
+            {submitting && <Spinner size="large" color="teal" />}
+
+            {errors.length > 0 && (
+              <Banner status="critical">
+                <ul>
+                  {errors.map(error => (
+                    <li key={error.message + error.field}>{error.message}</li>
+                  ))}
+                </ul>
+              </Banner>
+            )}
+
+            <FormLayout>
+              <Textfield label="Email" {...fields.email} />
+              <Textfield label="Password" {...fields.password} />
+            </FormLayout>
+
+            <Button type="submit" disabled={submitting}>
+              Save
+            </Button>
+          </form>
+        );
+      }}
+    </FormState>
+  );
+}
+```
+
+### Dirty Tracking
+
+Often you want to change some UI behaviour based on whether a form has changed form it's initial values. `<FormState />` supports this both on the form and individual field level.
+
+```typescript
+// MyPage.ts
+import {TextField, Button} from '@shopify/polaris';
+import FormState from '@shopify/react-form-state';
+
+export function MyPage() {
+  return (
+    <FormState
+      initialValues={{email: '', password: ''}}
+    >
+      {({fields, dirty}) => {
+        return (
+          <form>
+            <FormLayout>
+              <Textfield label="Email" {...fields.email} />
+              <Textfield label="Password" {...fields.password} />
+              {fields.pasword.dirty && <p>you have editied this</p>}
```

Comment:
can we make this a temporary constant outside of the return value? Also, `editied` -> `edited`

Reply:
ya sure